pkg/expr/expr.go

```diff
@@ -0,0 +1,365 @@
+	token.SUB: "SUB",
+	token.MUL: "MUL",
+	token.QUO: "QUO",
+	token.REM: "REM",
+
+	token.AND:  "AND",
+	token.OR:   "OR",
+	token.XOR:  "XOR",
+	token.LAND: "LAND",
+	token.LOR:  "LOR",
+
+	token.EQL: "EQ",
+	token.LSS: "LT",
+	token.GTR: "GT",
+	token.NOT: "NOT",
+
+	token.NEQ: "NEQ",
+	token.LEQ: "LEQ",
+	token.GEQ: "GEQ",
+
+	token.LBRACK: "INDEX",
+}
+
+var typeMap = map[token.Token]config.ValueType{
+	token.INT:    config.INT64,
+	token.FLOAT:  config.DOUBLE,
+	token.CHAR:   config.STRING,
+	token.STRING: config.STRING,
+}
+
+// Expression is a simplified expression AST
+type Expression struct {
+	// Oneof the following
+	Const *Constant
+	Var   *Variable
+	Fn    *Function
+}
+
+// Eval evaluates the expression given an attribute bag and a function map.
+func (e *Expression) Eval(attrs attribute.Bag, fMap map[string]Func) (interface{}, error) {
+	if e.Const != nil {
+		return e.Const.TypedValue, nil
+	}
+	if e.Var != nil {
+		v, ok := attribute.Value(attrs, e.Var.Name)
+		if !ok {
+			return nil, fmt.Errorf("unresolved attribute %s", e.Var.Name)
+		}
+		return v, nil
+	}
+	// may panic
+	return e.Fn.Eval(attrs, fMap)
+}
+
+// String produces postfix version with all operators converted to function names
+func (e *Expression) String() string {
+	if e.Const != nil {
+		return e.Const.String()
+	}
+	if e.Var != nil {
+		return e.Var.String()
+	}
+	if e.Fn != nil {
+		return e.Fn.String()
+	}
+	return "<nil>"
+}
+
+// newConstant creates a new constant of given type.
+// It also stores a typed form of the constant.
+func newConstant(v string, vType config.ValueType) (*Constant, error) {
+	var typedVal interface{}
+	var err error
+	switch vType {
+	case config.INT64:
+		if typedVal, err = strconv.ParseInt(v, 10, 64); err != nil {
+			return nil, err
+		}
+	case config.DOUBLE:
+		if typedVal, err = strconv.ParseFloat(v, 64); err != nil {
+			return nil, err
+		}
+	default: // string
+		if typedVal, err = strconv.Unquote(v); err != nil {
+			return nil, err
+		}
+	}
+	return &Constant{Value: v, Kind: vType, TypedValue: typedVal}, nil
+}
+
+// Constant models a typed constant.
+type Constant struct {
+	Value      string
+	TypedValue interface{}
+	Kind       config.ValueType
+}
+
+func (c *Constant) String() string {
+	return c.Value
+}
+
+// Variable models a variable.
+type Variable struct {
+	Name string
+}
+
+func (v *Variable) String() string {
+	return "$" + v.Name
+}
+
+// Function models a function with multiple parameters
+// 1st arg can be thought of as the receiver.
+type Function struct {
+	Name string
+	Args []*Expression
+}
+
+func (f *Function) String() string {
+	var w bytes.Buffer
+	w.WriteString(f.Name + "(")
+	for idx, arg := range f.Args {
+		if idx != 0 {
+			w.WriteString(", ")
+		}
+		w.WriteString(arg.String())
+	}
+	w.WriteString(")")
+	return w.String()
+}
+
+// Eval evaluate function.
+func (f *Function) Eval(attrs attribute.Bag, fMap map[string]Func) (interface{}, error) {
+	fn := fMap[f.Name]
+	if fn == nil {
```

Comment:
fn, found := fMap[f.Name]
    if !found {
        ...
    }

is more idiomatic